### ISSUE TITLE
What is the purpose of  "attn_weights += attn_mask" 

### ISSUE THREAD

**Issue by SamChen** [pytorch/fairseq#2963]
## ❓ Questions and Help
I was reading the code about multihead attention. I could not figure out the intension of this line.
https://github.com/pytorch/fairseq/blob/dea66cc294a18dd4d9e59aa0af8d51f951e83884/fairseq/modules/multihead_attention.py#L335

I guess it intend to mask all irrelevant scores/attn_weights generated by the Keys and Querys operation. But, the addition operator makes me feel confusion. I have two guess:

1.  The mask is a binary tensor. Each element is either 1 or 0. If this is the case, then, the addition does not make sense at all. Because, in this way, it still assign some attention to those unwanted Values.

2. The mask is still a binary tensor. But, each element is either the largest value of uint8 or 0. This way, it is actually utilize softmax' shift invariance property to minimize the attention assigned to unwanted Values. 

I tried  to find an example use case in fairseq's model examples. But, maybe it is because I am not familiar with the source code organization, I have not found any yet. 

Can someone explain this to me? OR, can someone shared me a link that include an example about the standard form of this attn_mask?

Thanks

**Comment by myleott** [comment]
We typically set it to -inf or similar negative value. For example: https://github.com/pytorch/fairseq/blob/1bc83c703ad70d7f62c1e54b197e29b95d07b1f0/fairseq/modules/transformer_layer.py#L121-L127
